### DISCUSSION TITLE
Learning Backtesting.py

### DISCUSSION THREAD

**Discussion by fchityan** [kernc/backtesting.py#1008] [General]
Hello everyone, 

I would like to learn how to make a custom indicators via talib (for example: Combined Bollinger Bands with ATR) using my own data. May I know where can I learn to code such custom indicators? I have search through Youtube and Google but nobody seems to be teaching such topic.

**Comment by peacerider** [comment]
I learned on the following pages.
https://greyhoundanalytics.com/blog/custom-indicators-in-backtestingpy/
Youtube explains in more detail.

**Reply by fchityan** [reply]
Yes, i have seen the video from the link you provided. 
However, i still feel that insufficient knowledge to create a custom indicators.

**Reply by peacerider** [reply]
The following pages may be useful.
https://medium.com/search?q=Backtesting.py

or this page is also useful for me. (but original language is Japanese)
https://myfrankblog.com/backtest_with_python/

**Reply by fchityan** [reply]
I have tried to code this Bollinger Bands strategy but there is no Sharpe ratio and the entry & exit timing is incorrect.
Please help to enlighten my issues!

    period = 20
    lower_std = -2
    upper_std = 2

    def init(self):
        close = self.data.Close
        self.upper_bound = self.I(talib.BBANDS, self.data.Close, self.period, self.upper_std)
        self.lower_bound = self.I(talib.BBANDS, self.data.Close, self.period, self.lower_std)

    def next(self):
        if self.data.Close > self.upper_bound:
            self.sell()
        elif self.data.Close < self.lower_bound:
            self.buy()

**Reply by peacerider** [reply]
Simple Example

```
from backtesting import Backtest, Strategy
import talib as ta
import datetime as dt
import pandas_datareader as pdr
from datetime import datetime

name = "APPLE"
code = "AAPL.US"

df = pdr.DataReader(code, "stooq").sort_index()
# df = df[dt.datetime(2000,1,1):dt.datetime.now()]


class myStratety(Strategy):
    period = 20
    lower_std = -2
    upper_std = 2
    ma_type = 0

    def init(self):
        close = self.data.Close
        self.upper_bound, _, self.lower_bound = \
            self.I(ta.BBANDS, self.data.Close, self.period, self.upper_std, self.upper_std, self.ma_type)

    def next(self):
        if self.data.Close > self.upper_bound:
            self.sell()
        elif self.data.Close < self.lower_bound:
            self.buy()

            
bt = Backtest(
    df,
    myStratety,
    cash=1000000,
    commission=.002,
    trade_on_close=False,
    exclusive_orders=True,
    hedging=False,
    margin=1
)
 
    
output = bt.run()
print(output)
bt.plot()
```

**Reply by fchityan** [reply]
I tried to import my own excel file into the code but the candlestick looks weird. 
data = pd.read_excel('/Users/Desktop/AAPL EOD Data.xlsx')
AAPL End of day excel file: [AAPL EOD Data.xlsx](https://github.com/kernc/backtesting.py/files/11925706/AAPL.EOD.Data.xlsx)


<img width="961" alt="Screenshot 2023-07-01 at 7 36 18 PM" src="https://github.com/kernc/backtesting.py/assets/54540183/bc9f1082-300d-4ac6-8f7e-cdd3dc5852f0">

**Reply by kernc** [reply]
> the candlestick looks weird

@cyfoo89 In your code, you have a buggy line such as:
```py
data['Open'] = data['High'] = data['Low'] = data['Close']
```

**Reply by peacerider** [reply]
I checked your data.
One of name of columns of Dataframe is wrong.
"Low "
space  is included.

**Reply by fchityan** [reply]
Sorry, i don understand what you mean by "Low " space is included.

**Reply by peacerider** [reply]
Sorry, my explanation is not enough.
I should have called "space" as "blank space".

Documents says,
"If any columns are missing, set them to what you have available, e.g. df['Open'] = df['High'] = df['Low'] = df['Close']"
https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#gsc.tab=0

**Reply by fchityan** [reply]
I have tried to input as what you have mention but it still the same. 

```
from backtesting import Backtest, Strategy
import pandas as pd
import talib as ta
import datetime as dt
import pandas_datareader as pdr
from datetime import datetime

data = pd.read_excel('/Users//AAPL EOD Data.xlsx')
data['Date'] = data['Date'].str.replace("T00:00:00+0000","")

data['Open'] = data['High'] = data['Low'] = data['Close']

class myStratety(Strategy):
    period = 20
    lower_std = -2
    upper_std = 2
    ma_type = 0

    def init(self):
        close = self.data.Close
        self.upper_bound, _, self.lower_bound = \
            self.I(ta.BBANDS, self.data.Close, self.period, self.upper_std, self.upper_std, self.ma_type)

    def next(self):
        if self.data.Close > self.upper_bound:
            self.sell()
        elif self.data.Close < self.lower_bound:
            self.buy()

            
bt = Backtest(
    data,
    myStratety,
    cash=1000000,
    commission=.002,
    trade_on_close=False,
    exclusive_orders=True,
    hedging=False,
    margin=1
)
 
    
output = bt.run()
print(output)
bt.plot()

```


<img width="939" alt="Screenshot 2023-07-04 at 9 39 19 PM" src="https://github.com/kernc/backtesting.py/assets/54540183/ae816c9d-9acd-44f7-9697-ad34f5eca028">

**Reply by peacerider** [reply]
You don' need following statement.
"data['Open'] = data['High'] = data['Low'] = data['Close']"

```
from backtesting import Backtest, Strategy
import talib as ta
import datetime as dt
import pandas_datareader as pdr
from datetime import datetime
import pandas as pd


data = pd.read_excel('AAPL_EOD_Data.xlsx').sort_index()
data = data.rename(columns={"Low ":"Low"})


class myStratety(Strategy):
    period = 20
    lower_std = -2
    upper_std = 2
    ma_type = 0

    def init(self):
        close = self.data.Close
        self.upper_bound, _, self.lower_bound = \
            self.I(ta.BBANDS, self.data.Close, self.period, self.upper_std, self.upper_std, self.ma_type)

    def next(self):
        if self.data.Close > self.upper_bound:
            self.sell()
        elif self.data.Close < self.lower_bound:
            self.buy()

            
bt = Backtest(
    data,
    myStratety,
    cash=1000000,
    commission=.002,
    trade_on_close=False,
    exclusive_orders=True,
    hedging=False,
    margin=1
)
 
    
output = bt.run()
print(output)
bt.plot()
```

**Reply by peacerider** [reply]
Column name "Low " needs to be changed to "Low".
*D1 cell of your Exel file is "Low ".  
This name is consisted by 4 character, "L", "o", "w", " " .
Last Blank Space " " is needed to be deleted.

**Reply by fchityan** [reply]
Thank you @peacerider for your help. Really appreciated.
